bindings/matrix-sdk-crypto-js/src/requests.rs

```diff
@@ -328,26 +325,73 @@ macro_rules! request {
+        (request_id = $request_id:expr, request = $request:expr)
+        $( extracts [ $( $field:ident: $field_type:ident ),* $(,)? ] )?
+        $( groups [ $( $grouped_field:ident ),* $(,)? ] )?
+    ) => {
+        {
+            Ok($destination_request {
+                id: $request_id,
+                $(
+                    $(
+                        $field: request!(@__field as $field_type (request = $request, field = $field)),
+                    )*
+                )?
+                $(
+                    body: {
+                        let mut map = serde_json::Map::new();
+                        $(
+                            map.insert(stringify!($grouped_field).to_owned(), serde_json::to_value(&$request.$grouped_field).unwrap());
+                        )*
+                        let object = serde_json::Value::Object(map);
+
+                        serde_json::to_string(&object)?.into()
+                    }
+                )?
+            })
+        }
+    };
+
+    ( @__field as json (request = $request:expr, field = $field:ident) ) => {
+        serde_json::to_string(&$request.$field)?.into()
+    };
+
+    ( @__field as string (request = $request:expr, field = $field:ident) ) => {
+        $request.$field.to_string().into()
+    };
+
+    ( @__field as duration (request = $request:expr, field = $field:ident) ) => {
+        $request.$field.map(|duration| duration.as_millis() as u64)
+    };
+
+    ( @__field as native (request = $request:expr, field = $field:ident) ) => {
+        $request.$field.clone()
+    };
 }
 
 // Outgoing Requests
-request!(KeysUploadRequest from OriginalKeysUploadRequest maps fields device_keys, one_time_keys, fallback_keys);
-request!(KeysQueryRequest from OriginalKeysQueryRequest maps fields timeout, device_keys, token);
-request!(KeysClaimRequest from OriginalKeysClaimRequest maps fields timeout, one_time_keys);
-request!(ToDeviceRequest from OriginalToDeviceRequest maps fields event_type, txn_id, messages);
-request!(SignatureUploadRequest from OriginalSignatureUploadRequest maps fields signed_keys);
-request!(RoomMessageRequest from OriginalRoomMessageRequest maps fields room_id, txn_id, content);
-request!(KeysBackupRequest from OriginalKeysBackupRequest maps fields rooms);
+request!(KeysUploadRequest from OriginalKeysUploadRequest groups fields device_keys, one_time_keys, fallback_keys);
+request!(KeysQueryRequest from OriginalKeysQueryRequest groups fields timeout, device_keys, token);
+request!(KeysClaimRequest from OriginalKeysClaimRequest groups fields timeout, one_time_keys);
+request!(ToDeviceRequest from OriginalToDeviceRequest extracts fields event_type: string, txn_id: string and groups fields messages);
+request!(SignatureUploadRequest from OriginalSignatureUploadRequest groups fields signed_keys);
+request!(RoomMessageRequest from OriginalRoomMessageRequest extracts fields room_id: string, txn_id: string and groups fields content);
```

Comment:
Extracting by looking up _inside_ a field is not supported now, but I'm working on it.

E.g, it's possible to extract one field (like `field_name`), but looking up inside a field (like `field_name.subfield_name`) is not supported for the moment.

Reply:
After an offline discussion, it appears that what we want is not accessing sub-field's values, but just having `content` as an extracted field rather than a grouped field. The content of `content` will be JSON-encoded as a string.